### PR TITLE
Quick-n-dirty error handling for file writes on 5.6 branch

### PR DESCRIPTION
**Pull request by ghoshkaj**
# Issue
https://github.com/Project-OSRM/osrm-backend/issues/3838 related to failures seen in production environments.

## Tasklist
 - [ ] throw errors to all file writes

### CODE REVIEW

File: include/storage/io.hpp [hunk 1]
```diff
@@ -253,7 +253,7 @@ class FileWriter
         const auto &result =
             output_stream.write(reinterpret_cast<const char *>(src), count * sizeof(T));
 
-        if (!result)
+        if (!result || !output_stream.good())
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: you can check `if (!stream)` everywhere

File: include/util/io.hpp [hunk 2]
```diff
@@ -36,6 +41,10 @@ bool serializeVector(std::ostream &stream, const std::vector<simple_type> &data)
     stream.write(reinterpret_cast<const char *>(&count), sizeof(count));
     if (!data.empty())
```
Comment (danpat) [inline include/util/io.hpp]: Missed one for the write on L41.

**Comment by danpat** [comment]
@ghoshkaj can you make the error text a bit clearer?  These messages are going to show up in production logs, so clear messages to whoever gets an alarm is always super helpful.  Something like:

  `Writing to .ebg (edge-based-graph) file failed`.

perhaps....

File: include/util/io.hpp (hunk 2 above)
Comment (ghoshkaj) [inline include/util/io.hpp]: No it's there. It's just on L44. It's referring to the same stream :)

**Comment by danpat** [comment]
@ghoshkaj If you can add a CHANGELOG.md ("Added error handling for file writes") entry, then I think this is basically ready to go for a 5.6.4 patch release.

File: include/util/indexed_data.hpp [hunk 3]
```diff
@@ -107,6 +108,11 @@ template <int N, typename T = std::string> struct VariableGroupBlock
 
         out.write((const char *)&refernce, sizeof(refernce));
 
+        if (!out.good())
+        {
+            throw util::exception(std::string("Error writing block reference: ") + SOURCE_REF);
+        }
```
Comment (daniel-j-h) [inline include/util/indexed_data.hpp]: re. `stream.good()` vs `stream.fail()` - if you check the table at the bottom 

http://en.cppreference.com/w/cpp/io/basic_ios/good

you will see that `fail` _is not_ the negation of `good`. It behaves differently for the eof bit. `.good` returns false if the eofbit is set, `.fail` returns false there too (!). I would keep this in mind and simply always check streams via `if (!stream)`.

Comment (ghoshkaj) [inline include/util/indexed_data.hpp]: Aha, okay, good an explanation of why one is better. Thanks @daniel-j-h !! 😄

**Pull request merged**